File: cpp/ctci/urlify.cc
#include <iostream>

using namespace std;

void urlify(char* in_str, int true_length) {
  // count spaces
  int spaces = 0;
  for (int i = 0; i < true_length; i++) {
    if (in_str[i] == ' ') {
      spaces += 1;
    }
  }
  // get new length
  int new_len = true_length + (2 * spaces);
  int j = new_len - 1;
  for (int i = true_length - 1; i >= 0; i--) {
    if (in_str[i] != ' ') {
      in_str[j] = in_str[i];
      j -= 1;
    } else {
      in_str[j] = '0';
      in_str[j - 1] = '2';
      in_str[j - 2] = '%';
      j -= 3;
    }
  }
}

int main() {
  char test [] = "Mr John Smith    ";
  int t_len = 13;
  cout << test << endl;
  urlify(test, t_len);
  cout << test << endl;
  return 0;
}

File: python/ctci/arrays_and_strings/urlify.py
def urlify(str, len):
    # count how many spaces
    space_cnt = 0
    for i in range(len):
        if str[i] == " ":
            space_cnt += 1
    # Figure out the new output string
    new_size = len + (2 * space_cnt)
    j = new_size - 1

    # In python str are immutable
    # make a list and then join them at the end
    out_str = [' '] * new_size
    for i in range(len - 1, -1, -1):
        if str[i] != " ":
            out_str[j] = str[i]
            j -= 1
        else: # str[i] is ' '
            out_str[j] = '0'
            out_str[j-1] = '2'
            out_str[j-2] = '%'
            j -= 3
    return ''.join(out_str)

if __name__ == '__main__':
    input_str = "Mr John Smith    "
    true_len = 13
    print(urlify(input_str, true_len))